### PR TITLE
keyvault: fixing the duplicate constant types

### PR DESCRIPTION
**Pull request by tombuildsstuff**
Fixes this:

```
$ autorest --tag=package-7.3 --go --verbose --use-onever --version=2.0.4421 --go.license-header=MICROSOFT_MIT_NO_VERSION --enum-prefix --openapi-type=data-plane --output-folder=../../sdk/keyvault/7.3/keyvault ../../config/key-vault/readme.md

FATAL: System.InvalidOperationException: Swagger document contains two or more x-ms-enum extensions with the same name 'ActionType' and different values: EmailContacts,AutoRenew vs. rotate,notify
```

### CODE REVIEW

File: specification/keyvault/data-plane/Microsoft.KeyVault/stable/7.3/certificates.json [hunk 1]
```diff
@@ -1630,7 +1630,7 @@
             "AutoRenew"
           ],
           "x-ms-enum": {
-            "name": "ActionType",
+            "name": "CertificateActionType",
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/stable/7.3/certificates.json:1633]: This is a breaking change. Why is this necessary, though? This name only affects SDKs, which should be referencing individual swaggers to generate separate SDKs for keys, secrets, certificates, and administration (basically, everything else). So conflicts between the two are irrelevant.

**Review by heaths — CHANGES_REQUESTED** [review]

File: specification/keyvault/data-plane/Microsoft.KeyVault/stable/7.3/certificates.json (hunk 1 above)
Comment (aristosvo) [reply specification/keyvault/data-plane/Microsoft.KeyVault/stable/7.3/certificates.json:1633]: `$.definitions.Actions` should be `$.definitions.Action`

```
  - from: certificates.json
    where: $.definitions.Action
    transform: $.properties.action_type["x-ms-enum"].name = "CertificateActionType";
  - from: keys.json
    where: $.definitions.LifetimeActionsType
    transform: $.properties.type["x-ms-enum"].name = "LifetimeActionType";
```

Comment (tombuildsstuff) [reply specification/keyvault/data-plane/Microsoft.KeyVault/stable/7.3/certificates.json:1633]: @heaths thanks for the suggestion.

> This is a breaking change for SDKs but shouldn't be necessary. This name only affects SDKs, which should be referencing individual swaggers to generate separate SDKs for keys, secrets, certificates, and administration (basically, everything else). So conflicts between the two are irrelevant unless you're generating a single SDK targeting all swaggers, though.

From our side we're trying to generate the SDK using the same approach that Track1 is using today, namely keeping this in a single package - so I think the autorest transform is _probably_ fine as a workaround here, but it begs the question of why not fix the data rather than applying a workaround?

From a logical deployment perspective, since these operations are exposed in the same Swagger directory, I'd expect them to be deployed together - so I'd argue that if these are intended to be separate SDKs then these should be split out (which would also remove the need for this workaround). Whilst this may seem like a pedantic point, fixing the data (either by merging this PR as-is, or by splitting these files into different directories) seems preferable to applying autorest workarounds, which should presumably only be a short-term solution and not widely used, else the source data isn't accurate?

> That begs the question, though: have you considered using our Go SDKs? https://github.com/Azure/azure-sdk-for-go/tree/main/sdk/keyvault

Unfortunately we're unable to use the Track2 SDK's for reasons we've raised with the Azure SDK team previously - instead [we're using `hashicorp/go-azure-sdk`](https://github.com/hashicorp/go-azure-sdk) which is generated from [`hashicorp/pandora`](https://github.com/hashicorp/pandora) (for Resource Manager) - and we'll be generating some Data Plane packages into another repository until we look into importing/generating those.

Comment (heaths) [reply specification/keyvault/data-plane/Microsoft.KeyVault/stable/7.3/certificates.json:1633]: There are a lot of services that don't use swagger in non-standard ways. Going forward with Cadl - a new description language that can generate both server- and client-side code (and swaggers, for legacy support) we'll create subdirectories per SDK, so these files won't be in the same directory. Ever since we refactored them, they were never meant to represent a single service.

Why can't you use the track 2? I've talked with several teams in the past and showed them our migration guides e.g., https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/keyvault/Azure.Security.KeyVault.Secrets/MigrationGuide.md. Everything should be supported 1:1 between track 1 and track 2 SDKs already except for SAS (not recommended anyway, which is why we didn't generate an SDK).

Comment (tombuildsstuff) [reply specification/keyvault/data-plane/Microsoft.KeyVault/stable/7.3/certificates.json:1633]: @heaths

>  It was a lot of extra work for almost every SDK language to trim support for endpoints that weren't supposed to be in an SDK, like keys and certs endpoints in the secrets SDK.

Potentially an odd take, but why would these be stripped out? If the Swagger documents them the SDK's should be generating them.

> Going forward with Cadl - a new description language that can generate both server- and client-side code (and swaggers, for legacy support) we'll create subdirectories per SDK, so these files won't be in the same directory. Ever since we refactored them, they were never meant to represent a single service.

Do you have any more context on Cadl / the timeframe there?

> Why can't you use the track 2?

We've previously given copious amounts of feedback to the Azure SDK Team as to why we're unable to use the Track2 Go SDK - and since we didn't hear back from the Azure SDK Team about our concerns - ultimately ended up having to build [`hashicorp/go-azure-sdk`](https://github.com/hashicorp/go-azure-sdk).

[The `hashicorp/go-azure-sdk` repository](https://github.com/hashicorp/go-azure-sdk) both fixes the concerns we had (albeit using the Track1 base layer as a temporary stop-gap, until we're fully off the Azure SDK for Go) and enables us to embed logic within the Go SDK to both make it more consistent (for example, [all operations take a Resource ID object](https://pkg.go.dev/github.com/hashicorp/go-azure-sdk@v0.20221024.1202112/resource-manager/compute/2021-11-01/virtualmachines#VirtualMachinesClient) where a Resource is being mutated) - which ultimately reduces the amount of code needed to interact with the SDK.

By comparison, adopting Track2 would mean:

* increasing the amount of code needed (both per-operation and in general)
* more admin overhead with nested Go modules (e.g. which semver relates to which API version? a large number of breaking changes every time we update a module)
* re-introducing the pain of composite Swaggers (which comes back to more admin overhead/which API version is _actually_ being used)
* hitting up against more Swagger inconsistencies (since `hashicorp/go-azure-sdk` corrects a bunch of these)

As such whilst Track2 may work great for folks using a small number of packages, unfortunately the design/architectural decisions mean it doesn't work for us since we're vendoring large amounts of the SDK (and are concerned about which API versions we're using) - as such this isn't something we're planning to adopt (even if the amount of code required to use Track2 was reduced in the same manner as `hashicorp/go-azure-sdk`).

Whilst that's unfortunate, in practice not using the Track2 SDK has enabled us to look into additional functionality (for example, Terraform Resource Generation) that wouldn't have been possible if we used the Track1 or Track2 SDKs (since we're normalizing the Swaggers, for example duck-typing into common types) - so whilst there is an overhead to generating this SDK ourselves, I'd argue it's working out really well for us - and as such unfortunately we have no plans to adopt Track2.

Comment (heaths) [reply specification/keyvault/data-plane/Microsoft.KeyVault/stable/7.3/certificates.json:1633]: The secrets package doesn't need support for keys or certificates endpoints, for example. We split packages up to be more area-focused, so having one giant file with all the endpoints didn't make sense. There are also numerous feature teams and refactoring the files made it easier for them to work and deploy in parallel as well.

We don't have a timeframe for when Key Vault will be migrated to Cadl, but you can find more information about Cadl here: https://github.com/microsoft/cadl

As of now, however, these swaggers have shipped this way for years and changing now is a breaking change. Please use the suggested (amended) autorest transforms.

**Comment by heaths** [comment]
We appreciate contributions to the Azure SDKs, but since this is an SDK-breaking change as noted above, we cannot accept this. If you're using autorest to generate your own SDK from all swaggers - which we split to better support the service and SDK feature teams with separate definitions - please use the [amended transforms above](https://github.com/Azure/azure-rest-api-specs/pull/21137#discussion_r996423099).

File: specification/keyvault/data-plane/Microsoft.KeyVault/stable/7.3/certificates.json (hunk 1 above)
Comment (tombuildsstuff) [reply specification/keyvault/data-plane/Microsoft.KeyVault/stable/7.3/certificates.json:1633]: @heaths 

Just to be clear, could you highlight _where_ this is a breaking change? Splitting this out is a breaking change, since the Azure SDK for Go has shipped this as a combined Swagger for many years:

* Create Certificate - https://github.com/Azure/azure-sdk-for-go/blob/438bb470af0f10f6952c59605ad8d7ef86acffba/services/keyvault/v7.1/keyvault/client.go#L367

* Create Key - https://github.com/Azure/azure-sdk-for-go/blob/438bb470af0f10f6952c59605ad8d7ef86acffba/services/keyvault/v7.1/keyvault/client.go#L462

* Set Secret - https://github.com/Azure/azure-sdk-for-go/blob/438bb470af0f10f6952c59605ad8d7ef86acffba/services/keyvault/v7.1/keyvault/client.go#L6092

As such I'm confused as to how this could be considered a breaking change? I can understand the desire to split this, fine, but I don't believe this is a breaking change.

Whilst we can look to add an AutoRest Transform to workaround this, my understanding is those are supposed to be a last resort - so I'm at a loss for why:

a) applying this change would be a breaking change (to either Track1, or Track2)
b) the recommended approach here is to apply a hack (an AutoRest Transform) rather than fix the underlying issue

Comment (tombuildsstuff) [reply specification/keyvault/data-plane/Microsoft.KeyVault/stable/7.3/certificates.json:1633]: any update @heaths?

Comment (tombuildsstuff) [reply specification/keyvault/data-plane/Microsoft.KeyVault/stable/7.3/certificates.json:1633]: Fundamentally the differentiation between Track1 and Track2 is irrelevant here though, this change wouldn't be a breaking change in either of them, so the reason for closing here is pretty misleading?

Comment (tombuildsstuff) [reply specification/keyvault/data-plane/Microsoft.KeyVault/stable/7.3/certificates.json:1633]: @heaths I get that you'd like us to use Track2, but I still think you've closed this for the wrong reasons here - @JeffreyRichter any chance you'd be able to take a look here? I think @heaths and I are at an impasse here?

Comment (JeffreyRichter) [reply specification/keyvault/data-plane/Microsoft.KeyVault/stable/7.3/certificates.json:1633]: I'll meet with Heath and discuss so I understand the issue and report back here (hopefully next week but it is Thanksgiving).

Comment (tombuildsstuff) [reply specification/keyvault/data-plane/Microsoft.KeyVault/stable/7.3/certificates.json:1633]: @JeffreyRichter did you have a chance to chat with Heath about this yet?

Comment (JeffreyRichter) [reply specification/keyvault/data-plane/Microsoft.KeyVault/stable/7.3/certificates.json:1633]: Yes, Heath & I spoke and as it is SDK breaking, we cannot accept this PR.

Comment (tombuildsstuff) [reply specification/keyvault/data-plane/Microsoft.KeyVault/stable/7.3/certificates.json:1633]: Any chance you could explain *how* this is SDK breaking? The Breaking Change Action above says this isn't a breaking change:

Screenshot 2022-12-12 at 10 01 44

**Comment by aristosvo** [comment]
@heaths @JeffreyRichter As both an Azure enthousiast and Terraform Provider contributor this annoys me a lot. Can we please find a way to work together in a constructive way and move this forward and out of the way?

Could one at least make sure that https://github.com/Azure/azure-rest-api-specs/tree/main/specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.4-preview.1 contains the proposed fix before someone freezes it?

**Comment by JeffreyRichter** [comment]
OK. Heath and I will discuss this again and see what we can do. Unfortunately, he is now out until after New Year's, so this will have to sit until we can connect. Sorry for the delay.

**Comment by jlichwa** [comment]
@aristosvo above is changing property name, which will not match the service implementation. This field is there since the beginning and change will break all customers. Proposed change is impossible. You can schedule a meeting with team to figure out other options (possibly fix track 2 to work as our recommended option). 

Key Vault PG team

**Comment by JeffreyRichter** [comment]
I found someone else to speak to about this and here's the deal...
For our Track 2 SDKs, we generate 1 package for certificates and a separate package for keys. This means that calling these 2 enums "ActionType" is not a problem because they are scoped to their respective packages.

We have already shipped these packages and the packages name the enum type from the swagger name.
Changing the enum name is breaking to us. The tooling is not catching this for some reason - I will contact the tooling team and get them to fix this bug.

If you are using our autorest tool to produce code from the swagger, you can use a swagger directive to rename the enum types for your code-gen. See https://github.com/Azure/autorest/blob/main/docs/generate/built-in-directives.md

**Comment by aristosvo** [comment]
@JeffreyRichter Thanks, but that leaves the terraform users now in the dark. The question is what we can do then to support probably the largest external user base of the Azure REST API in a proper way.

I opened up #21877 as that API version is still in preview, would that work?

**Comment by aristosvo** [comment]
> @aristosvo above is changing property name, which will not match the service implementation. This field is there since the beginning and change will break all customers. Proposed change is impossible. You can schedule a meeting with team to figure out other options (possibly fix track 2 to work as our recommended option).
> 
> Key Vault PG team

I don't think that is the case. This are some annotations which influence the output of the SDK generation process, not the calls towards the API implementation AFAIK.

I understand that an breaking change in an SDK is annoying as well, from a Terraform Provider perspective we've dealt with these for years in Track1. Dropping proper Track1 support without an easy transition to Track2 is a similar annoying move.

To be clear, I don't want to complain and put you on the spot or anything like that, just want to make sure we're making some moves forward for the users of the Terraform Provider for Azure, which have waited long enough already.

**Comment by jlichwa** [comment]
@aristosvo it is not annotation. It is an enum which is passed to service and it will break all clients. Track 1 is deprecated and it is not supported for any new features. You will need to migrate/create new Track 2 SDK. 
There are no changes allowed in current not subsequent version, completely new endpoint would have to be introduced and old one would need to support for at least 3 years in subsequent versions. There is no plan to do that since we migrated to Track 2 and Track 1 is deprecated and support will end in March -
 https://azure.github.io/azure-sdk/releases/deprecated/index.html

**Comment by jlichwa** [comment]
@aristosvo I discussed with @JeffreyRichter about various options. The path forward would be to have a meeting with me and @heaths when he comes back next year to discuss it.

**Comment by JeffreyRichter** [comment]
Yes Tom, what you propose is possible. And we did discuss this we'd have to coordinate it across all the SDK language owners which is tedious but possible. Jack & I also discussed some other alternatives too, but I prefer to let him tell you about them when you & he meet.

**Comment by tombuildsstuff** [comment]
hey @JeffreyRichter - just wondering if you'd had a chance to sync up with Jack/Heath yet on how's best to proceed here?

**Comment by JeffreyRichter** [comment]
Jack & I spoke last month. The ball is in @jlichwa's court to talk to you.

**Comment by jlichwa** [comment]
The best way you to contact our team is to create support request as I mentioned above in comments. Looking forward to the meeting.

**Comment by jlichwa** [comment]
@tombuildsstuff  7.4 with fix is now available : https://github.com/Azure/azure-rest-api-specs/tree/main/specification/keyvault/data-plane/Microsoft.KeyVault/stable/7.4
